webrender/src/prim_store.rs

```diff
@@ -0,0 +1,924 @@
+pub struct SpecificPrimitiveIndex(pub usize);
+
+impl SpecificPrimitiveIndex {
+    pub fn invalid() -> SpecificPrimitiveIndex {
+        SpecificPrimitiveIndex(usize::MAX)
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct PrimitiveIndex(pub usize);
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum PrimitiveKind {
+    Rectangle,
+    TextRun,
+    Image,
+    Border,
+    Gradient,
+    BoxShadow,
+}
+
+#[derive(Debug, Clone)]
+pub struct PrimitiveGeometry {
+    pub local_rect: Rect<f32>,
+    pub local_clip_rect: Rect<f32>,
+}
+
+// TODO(gw): Pack the fields here better!
+#[derive(Debug)]
+pub struct PrimitiveMetadata {
+    pub is_opaque: bool,
+    pub need_to_build_cache: bool,
+    pub color_texture_id: TextureId,
+    pub clip_index: Option<GpuStoreAddress>,
+    pub prim_kind: PrimitiveKind,
+    pub cpu_prim_index: SpecificPrimitiveIndex,
+    pub gpu_prim_index: GpuStoreAddress,
+    pub gpu_data_address: GpuStoreAddress,
+    pub gpu_data_count: i32,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum TextureCoordKind {
+    Normalized = 0,
+    Pixel,
+}
+
+#[derive(Debug, Clone)]
+pub struct RectanglePrimitive {
+    pub color: ColorF,
+}
+
+#[derive(Debug)]
+pub enum ImagePrimitiveKind {
+    Image(ImageKey, ImageRendering, Size2D<f32>, Size2D<f32>),
+    WebGL(WebGLContextId),
+}
+
+#[derive(Debug)]
+pub struct ImagePrimitiveCpu {
+    pub kind: ImagePrimitiveKind,
+}
+
+#[derive(Debug, Clone)]
+pub struct ImagePrimitiveGpu {
+    pub uv0: Point2D<f32>,
+    pub uv1: Point2D<f32>,
+    pub stretch_size: Size2D<f32>,
+    pub tile_spacing: Size2D<f32>,
+}
+
+#[derive(Debug, Clone)]
+pub struct BorderPrimitiveCpu {
+    pub inner_rect: Rect<f32>,
+}
+
+#[derive(Debug, Clone)]
+pub struct BorderPrimitiveGpu {
+    pub style: [f32; 4],
+    pub widths: [f32; 4],
+    pub colors: [ColorF; 4],
+    pub radii: [Size2D<f32>; 4],
+}
+
+#[derive(Debug, Clone)]
+pub struct BoxShadowPrimitive {
+    pub src_rect: Rect<f32>,
+    pub bs_rect: Rect<f32>,
+    pub color: ColorF,
+    pub border_radii: Point2D<f32>,
+    pub blur_radius: f32,
+    pub inverted: f32,
+}
+
+#[repr(u32)]
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum GradientType {
+    Horizontal,
+    Vertical,
+    Rotated,
+}
+
+#[derive(Debug, Clone)]
+pub struct GradientStop {
+    pub color: ColorF,
+    pub offset: f32,
+    pub padding: [f32; 3],
+}
+
+#[derive(Debug, Clone)]
+pub struct GradientPrimitiveGpu {
+    pub start_point: Point2D<f32>,
+    pub end_point: Point2D<f32>,
+    pub kind: f32,
+    pub padding: [f32; 3],
+}
+
+#[derive(Debug)]
+pub struct GradientPrimitiveCpu {
+    pub stops_range: ItemRange,
+    pub kind: GradientType,
+    pub reverse_stops: bool,
+}
+
+#[derive(Debug, Clone)]
+pub struct InstanceRect {
+    pub rect: Rect<f32>,
+}
+
+#[derive(Debug, Clone)]
+pub struct TextRunPrimitiveGpu {
+    pub color: ColorF,
+}
+
+#[derive(Debug, Clone)]
+pub struct TextRunPrimitiveCpu {
```

Comment:
Structs get split SoA style into portions that get directly pushed into a vertex texture (the gpu version) and cpu-side only portions (typically used to look up resource lists etc).